Core/Src/ws2812.c

```diff
@@ -0,0 +1,106 @@
+/*
+ * ws2812.c
+ *
+ *  Created on: Sep 22, 2024
+ *      Author: Anthony Luo and Fola Fatola
+ *
+ *      Reference:
+ *      https://www.thevfdcollective.com/blog/stm32-and-sk6812-rgbw-led
+ */
+
+#include "stm32l4xx_hal.h"
+#include "ws2812.h"
+#include "hex_to_rgb_conversion.h"
+
+extern TIM_HandleTypeDef htim1;
+extern DMA_HandleTypeDef hdma_tim1_ch2;
+
+uint8_t out_buf[BUFF_SIZE]; // We add 24 to the end so that we have 24 (one LED) bits of 0
+
+
+void led_buffer_init() {
+	for (int i = 0; i < BUFF_SIZE; i++) {
+		//First and last 24 bytes LEDs get no valuable data so there's a gap between
+		//data transfers.
+		//For everything else, the buffer is filled with PWM_LO (for 0 brightness)
+		//by default.
+		if (i < 24 || i > (NUM_LEDS + 1) * 24) {
+			out_buf[i] = 0;
+		} else {
+			out_buf[i] = PWM_LO;
+		}
+	}
+}
+
+//Writes the color specified by "color_code" to the Neopixel specified by the index.
+void led_set_color(uint8_t index, uint32_t color_code) {
+
+	//Get the brightness of each color
+	uint8_t green_brightness_byte = get_green_byte_from_hex(color_code);
+	uint8_t red_brightness_byte = get_red_byte_from_hex(color_code);
+	uint8_t blue_brightness_byte = get_blue_byte_from_hex(color_code);
+
+	led_set_RGB_index(index, green_brightness_byte, red_brightness_byte, blue_brightness_byte, color_code);
+}
+
+void led_set_RGB_index(uint8_t index, uint8_t green_byte, uint8_t red_byte, uint8_t blue_byte, uint32_t color_code) {
+
+	uint8_t mask = 1;
+	//For each color's brightness byte, mask off each bit 1 at a time.
+	//If the color bit is a 1, write PWM_high to the appropriate index of the out_buf.
+	//If the color bit is 0, write PWM_low.
+	//Less significant bits map to higher indices in the array due to Neopixel frame layout.
+	//Please look at "Composition of 24 bit data" in https://cdn-shop.adafruit.com/datasheets/WS2812B.pdf
```

Comment:
This is definitely a NIT, but we strongly prefer downloading the PDF of the datasheet and uploading it to our own internal confluence, that way if the external site goes down the link is not broken.